### PR TITLE
enh(conf) add DeleteHostCategory endpoint

### PR DESCRIPTION
**Pull request by a-launois**
## Description

MON-16037

Create API endpoint : DELETE /configuration/hosts/categories/{id}

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 21.10.x
- [ ] 22.04.x
- [ ] 22.10.x
- [x] 23.04.x (master)

#### Community contributors & Centreon team

- [x] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

[MON-16037]: https://centreon.atlassian.net/browse/MON-16037?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: centreon/src/Core/HostCategory/Infrastructure/API/DeleteHostCategory/DeleteHostCategoryRoute.yaml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+deleteHostCategory:
+  methods: DELETE
+  path: /configuration/hosts/categories/{hostCategoryId}
+  requirements:
+    hostCategoryId: '\d+'
+  controller: 'Core\HostCategory\Infrastructure\API\DeleteHostCategory\DeleteHostCategoryController'
+  condition: "request.attributes.get('version') >= 22.10"
```
Comment (a-launois) [inline centreon/src/Core/HostCategory/Infrastructure/API/DeleteHostCategory/DeleteHostCategoryRoute.yaml:7]:
```suggestion
  condition: "request.attributes.get('version') >= 23.04"
```

**Review by jeremyjaouen — APPROVED** [review]

File: centreon/src/Core/HostCategory/Application/UseCase/DeleteHostCategory/DeleteHostCategory.php [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+                        'accessgroups' => $accessGroups
+                    ]);
+                    $presenter->setResponseStatus(new NotFoundResponse('Host category'));
+                }
+            } else {
+                $this->error('User doesn\'t have sufficient rights to see host category', [
+                    'user_id' => $this->user->getId(),
+                ]);
+                $presenter->setResponseStatus(
+                    new ForbiddenResponse('You are not allowed to delete host categories')
```
Comment (abuathier) [inline centreon/src/Core/HostCategory/Application/UseCase/DeleteHostCategory/DeleteHostCategory.php:81]: The message needs to be in a `HostCategoryException` class.
⚠️ update the test.

File: centreon/src/Core/HostCategory/Application/UseCase/DeleteHostCategory/DeleteHostCategory.php [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+                );
+            }
+        } catch (\Throwable $ex) {
+            $presenter->setResponseStatus(new ErrorResponse('Error while deleting host category'));
```
Comment (abuathier) [inline centreon/src/Core/HostCategory/Application/UseCase/DeleteHostCategory/DeleteHostCategory.php:85]: The message needs to be in a `HostCategoryException` class.
⚠️ update the test.

File: centreon/src/Core/HostCategory/Infrastructure/API/DeleteHostCategory/DeleteHostCategoryController.php [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+use Core\HostCategory\Application\UseCase\DeleteHostCategory\DeleteHostCategory;
+use Core\Infrastructure\Common\Api\DefaultPresenter;
+
+final class DeleteHostCategoryController extends AbstractController
+{
+    public function __invoke(
+        int $hostCategoryId,
+        DeleteHostCategory $useCase,
+        DefaultPresenter $presenter
+    ): object {
```
Comment (abuathier) [inline centreon/src/Core/HostCategory/Infrastructure/API/DeleteHostCategory/DeleteHostCategoryController.php:36]: Return type hinting more precise ? `Response`

File: centreon/src/Core/HostCategory/Infrastructure/API/DeleteHostCategory/DeleteHostCategoryRoute.yaml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+deleteHostCategory:
```
Comment (abuathier) [inline centreon/src/Core/HostCategory/Infrastructure/API/DeleteHostCategory/DeleteHostCategoryRoute.yaml:1]: `PascalCase`

File: centreon/tests/php/Core/HostCategory/Application/UseCase/DeleteHostCategory/DeleteHostCategoryPresenterStub.php [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+declare(strict_types=1);
+
+namespace Tests\Core\HostCategory\Application\UseCase\DeleteHostCategory;
+
+use Symfony\Component\HttpFoundation\Response;
+use Core\Application\Common\UseCase\AbstractPresenter;
+use Core\HostCategory\Application\UseCase\DeleteHostCategory\DeleteHostCategoryResponse;
+
+class DeleteHostCategoryPresenterStub extends AbstractPresenter
```
Comment (abuathier) [inline centreon/tests/php/Core/HostCategory/Application/UseCase/DeleteHostCategory/DeleteHostCategoryPresenterStub.php:30]: Maybe use directly the `DefaultPresenter` : it works for me.

File: centreon/src/Core/HostCategory/Infrastructure/API/DeleteHostCategory/DeleteHostCategoryController.php (hunk 4 above)
Comment (a-launois) [reply centreon/src/Core/HostCategory/Infrastructure/API/DeleteHostCategory/DeleteHostCategoryController.php:36]: As existing controller seems to return object as well, I left it like that for now

File: centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php [hunk 7]
```diff
@@ -198,4 +198,71 @@ private function hasRestrictedAccessToHostCategories(array $accessGroupIds): boo
...
+
+    /**
+     * @inheritDoc
+     */
+    public function exists(int $hostCategoryId): bool
+    {
+        $this->info('Check existance of host category with id #' . $hostCategoryId);
+
+        $request = $this->translateDbName(
+            'SELECT hc.hc_id FROM `:db`.hostcategories hc WHERE hc.hc_id = :hostCategoryId'
```
Comment (abuathier) [inline centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php:210]: As told by Laurent on my own PRs, there a _relative_ risk around the `hc_id=0` which will be cast to `false`.
Prefer then the use of `SELECT 1` which is `hc_id` agnostic.
```suggestion
            'SELECT 1 FROM `:db`.hostcategories hc WHERE hc.hc_id = :hostCategoryId'
```

File: centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php [hunk 8]
```diff
@@ -198,4 +198,71 @@ private function hasRestrictedAccessToHostCategories(array $accessGroupIds): boo
...
+
+        if (empty($accessGroups)) {
+            $this->debug('Access groups array empty');
+            return false;
+        }
+
+        $concat = new SqlConcatenator();
+
+        $accessGroupIds = array_map(
+            fn($accessGroup) => $accessGroup->getId(),
```
Comment (abuathier) [inline centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php:237]: Please check the code style on this file 🙏🏻 with php-cs-fixer
```suggestion
            fn ($accessGroup) => $accessGroup->getId(),
```

File: centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php [hunk 9]
```diff
@@ -198,4 +198,71 @@ private function hasRestrictedAccessToHostCategories(array $accessGroupIds): boo
...
+            $accessGroups
+        );
+
+        $request = $this->translateDbName(
+            'SELECT hc.hc_id
```
Comment (abuathier) [inline centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php:242]: Same remark as previous `SELECT 1`.
```suggestion
            'SELECT 1
```

**Review by abuathier — APPROVED** [review]

File: centreon/lang/fr_FR.UTF-8/LC_MESSAGES/messages.po [hunk 10]
```diff
@@ -17220,3 +17220,12 @@ msgstr "Vous n'êtes pas autorisé à accéder aux catégories d'hôtes"
 
 msgid "Error while searching for host categories"
 msgstr "Erreur lors de la recherche des catégories d'hôtes"
+
+msgid "You are not allowed to delete host categories"
+msgstr "Vous n'êtes pas autorisé à supprimer des catégories d'hôtes"
+
+msgid "Host category not found"
+msgstr "Catégorie d'hôte introuvable"
```
Comment (cg-tw) [inline centreon/lang/fr_FR.UTF-8/LC_MESSAGES/messages.po:17228]:
```suggestion
msgstr "Catégorie d'hôtes introuvable"
```

File: centreon/lang/fr_FR.UTF-8/LC_MESSAGES/messages.po [hunk 11]
```diff
@@ -17220,3 +17220,12 @@ msgstr "Vous n'êtes pas autorisé à accéder aux catégories d'hôtes"
...
+
+msgid "Error while deleting host category"
+msgstr "Erreur lors de la suppression d'une catégorie d'hôte"
```
Comment (cg-tw) [inline centreon/lang/fr_FR.UTF-8/LC_MESSAGES/messages.po:17231]:
```suggestion
msgstr "Erreur lors de la suppression d'une catégorie d'hôtes"
```

File: centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php [hunk 12]
```diff
@@ -107,6 +108,73 @@ public function findAllByAccessGroups(array $accessGroups, ?RequestParametersInt
         return $this->retrieveHostCategories($concatenator, $requestParameters);
     }
 
+    /**
+     * @inheritDoc
+     */
+    public function exists(int $hostCategoryId): bool
+    {
+        $this->info('Check existance of host category with id #' . $hostCategoryId);
```
Comment (cg-tw) [inline centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php:116]:
```suggestion
        $this->info('Check existence of host category with id #' . $hostCategoryId);
```

File: centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php [hunk 13]
```diff
@@ -107,6 +108,73 @@ public function findAllByAccessGroups(array $accessGroups, ?RequestParametersInt
...
+        return (bool) $statement->fetchColumn();
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function existsByAccessGroups(int $hostCategoryId, array $accessGroups): bool
+    {
+        $this->info(
+            'Check existance of host category by access groups',
```
Comment (cg-tw) [inline centreon/src/Core/HostCategory/Infrastructure/Repository/DbReadHostCategoryRepository.php:134]:
```suggestion
            'Check existence of host category by access groups',
```

File: centreon/tests/php/Core/HostCategory/Application/UseCase/DeleteHostCategory/DeleteHostCategoryTest.php [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+
+    $useCase($this->hostCategoryId, $this->presenter);
+
+    expect($this->presenter->getResponseStatus())
+        ->toBeInstanceOf(ErrorResponse::class)
+        ->and($this->presenter->getResponseStatus()->getMessage())
+        ->toBe(HostCategoryException::deleteHostCategory(new \Exception())->getMessage());
+});
+
+it('should present an ForbiddenResponse when a non-admin user has unsufficient rights', function (): void {
```
Comment (cg-tw) [inline centreon/tests/php/Core/HostCategory/Application/UseCase/DeleteHostCategory/DeleteHostCategoryTest.php:80]:
```suggestion
it('should present a ForbiddenResponse when a non-admin user has insufficient rights', function (): void {
```

**Review by cg-tw — APPROVED** [review]

**Pull request merged**